src/flp/szk.rs

```diff
@@ -209,17 +207,54 @@ impl<F: FieldElement + Decode, const SEED_SIZE: usize> ParameterizedDecode<(bool
 }
 
 /// A tuple containing the state and messages produced by an SZK query.
-#[cfg(test)]
 #[derive(Clone, Debug)]
 pub struct SzkQueryShare<F: FieldElement, const SEED_SIZE: usize> {
     joint_rand_part_opt: Option<Seed<SEED_SIZE>>,
-    flp_verifier: Vec<F>,
+    pub(crate) flp_verifier: Vec<F>,
 }
 
-/// The state that needs to be stored by an Szk verifier between query() and decide()
-pub type SzkQueryState<const SEED_SIZE: usize> = Option<Seed<SEED_SIZE>>;
+impl<F: FieldElement, const SEED_SIZE: usize> Encode for SzkQueryShare<F, SEED_SIZE> {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        if let Some(ref part) = self.joint_rand_part_opt {
+            part.encode(bytes)?
+        };
+
+        encode_fieldvec(&self.flp_verifier, bytes)?;
+        Ok(())
+    }
+
+    fn encoded_len(&self) -> Option<usize> {
+        Some(
+            self.flp_verifier.len() * F::ENCODED_SIZE
+                + match self.joint_rand_part_opt {
+                    Some(ref part) => part.encoded_len()?,
+                    None => 0,
+                },
+        )
+    }
+}
+
+impl<F: FieldElement + Decode, const SEED_SIZE: usize> ParameterizedDecode<(bool, usize)>
+    for SzkQueryShare<F, SEED_SIZE>
+{
+    fn decode_with_param(
+        (requires_joint_rand, verifier_len): &(bool, usize),
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        if *requires_joint_rand {
+            Ok(SzkQueryShare {
+                joint_rand_part_opt: Some(Seed::<SEED_SIZE>::decode(bytes)?),
+                flp_verifier: decode_fieldvec(*verifier_len, bytes)?,
+            })
+        } else {
+            Ok(SzkQueryShare {
+                joint_rand_part_opt: None,
+                flp_verifier: decode_fieldvec(*verifier_len, bytes)?,
+            })
+        }
```

Comment:
Don't repeat code if you don't need to.
```suggestion
        Ok(SzkQueryShare {
            joint_rand_part_opt: (*requires_joint_rand)
                .then(|| Seed::<SEED_SIZE>::decode(bytes))
                .transpose()?,
            flp_verifier: decode_fieldvec(*verifier_len, bytes)?,
        })
```